### PR TITLE
bpo-34720: Fix test_importlib.test_bad_traverse for AIX

### PR DESCRIPTION
**Pull request by aixtools**
Fix  Modules/_testmultiphase.c so that it exits with non-zero status on AIX just as other systems do (non zero exit status, e.g. as result of a segmentation fault) when a NULL pointer is accessed for data.

https://bugs.python.org/issue34720

### CODE REVIEW

**Comment by aixtools** [comment]
close, and reopen, to rerun the tests (still showing one red X atm).

File: Misc/NEWS.d/next/Tests/2018-09-18-12-03-41.bpo-34720.d5uOoP.rst
Comment (taleinat) [inline Misc/NEWS.d/next/Tests/2018-09-18-12-03-41.bpo-34720.d5uOoP.rst]: Test-only changes usually don't require a NEWS entry. In this case, it should be removed.

(Bedevere will complain, but you may ignore that, and I or another core dev will simply add the "skip news" tag.)

File: Modules/_testmultiphase.c
Comment (taleinat) [inline Modules/_testmultiphase.c]: Perhaps this could replace the `Py_VISIT(m_state->integer);` call on all platforms, thus avoiding the `#ifdef`?

Comment (serhiy-storchaka) [reply Modules/_testmultiphase.c]: Calling `Py_VISIT(m_state->integer)` is the purpose of this test.

But I wondering if it could be better to just skip the test on AIX in Python code.

File: Misc/NEWS.d/next/Tests/2018-09-18-12-03-41.bpo-34720.d5uOoP.rst
Comment (aixtools) [reply Misc/NEWS.d/next/Tests/2018-09-18-12-03-41.bpo-34720.d5uOoP.rst]: I'll remove the News, as requested.

File: Modules/_testmultiphase.c
Comment (aixtools) [reply Modules/_testmultiphase.c]: Obviously, I cannot "decide" what is "cleaner" - letting the test run, but have a modification in the C code the test calls, or having an "SkipIf" in the test (and move the comment to the test, so it is clear why the skip is needed).

I will make the recommended change.

Comment (ncoghlan) [reply Modules/_testmultiphase.c]: This test case originated from https://github.com/python/cpython/commit/c2b0b12d1a137ada1023ab7c10b8d9a0249d95f9, and the new behaviour it's actually attempting to verify is:

1. In debug mode, module creation automatically invokes the registered traverse function (for eager detection of problematic traversal function definitions)
2. When the traversal function isn't checking correctly for a valid module state, the module creation fails

Actually implementing a valid traversal function here isn't overly important, as the module using it never actually gets created.

So I think @taleinat's right that this bad traversal function can be simplified to avoid the `#ifdef`, but I think it can be simplified all the way to being just:

    m_state = PyModule_GetState(self);
    /* The following assertion mimics any traversal function that doesn't correctly handle
     * the case during module creation where the module state hasn't been created yet.
     */
    assert(m_state != NULL);
    Py_VISIT(m_state->integer);
    return 0;

That way the case of interest will always fail on the assert() call, and never even attempt to dereference the pointer.

**Review by ncoghlan — CHANGES_REQUESTED** [review]
I think we can make this test more robust in general by replacing the null-pointer dereference with a known-to-be-incorrect `assert()` call.

**Comment by ncoghlan** [comment]
Note that the failing CI check is just looking for a NEWS entry.

The entry for this one would go in the "Tests" section, and can be generated with the Blurb CLI or the web service at https://blurb-it.herokuapp.com/ (as per the updated https://devguide.python.org/committing/#what-s-new-and-news-entries section)

**Comment by aixtools** [comment]
I have made the requested changes; please review again.

**Review by ncoghlan — APPROVED** [review]

**Comment by ncoghlan** [comment]
Since I had forgotten why we didn't need to worry about the `assert()` getting compiled out in fully optimised release builds, I modified the block comment in the test traversal function to explain the details of that.

**Comment by ncoghlan** [comment]
Miss Islington will merge this automatically once CI completes successfully, but after that we'll want to keep an eye on https://buildbot.python.org/all/#/grid?branch=master&tag=stable&tag=nondebug to make sure release mode builds aren't unexpectedly affected by the change.

**Pull request merged**
